src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,334 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock\Store;
+
+use MongoDB\BSON\UTCDateTime;
+use MongoDB\Client;
+use MongoDB\Collection;
+use MongoDB\Driver\Exception\WriteException;
+use MongoDB\Exception\DriverRuntimeException;
+use MongoDB\Exception\InvalidArgumentException as MongoInvalidArgumentException;
+use MongoDB\Exception\UnsupportedException;
+use Symfony\Component\Lock\Exception\InvalidArgumentException;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\LockExpiredException;
+use Symfony\Component\Lock\Exception\LockStorageException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * MongoDbStore is a StoreInterface implementation using MongoDB as a storage
+ * engine.
+ *
+ * @author Joe Bennett <joe@assimtech.com>
+ */
+class MongoDbStore implements StoreInterface
```

Comment:
Starting a thread here:

> > Thanks for your patience and answering all my questions @kralos
> > The only concern that I have left is, this implementation leaves with a possibility of leaving behind stale locks in the DB if an application encounters a crash or some other failure or if the version of mongodb is older than `2.2`. This can be bad if you have a high throughput application that requires a lock only once to avoid duplicate processing or something.
> > I am thinking maybe it should also do something similar to `PDOStore` which seems to clean up stale locks at random intervals
> > https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Lock/Store/PdoStore.php#L143
> 
> I've had a look at `PdoStore`'s implementation and I have some concerns with it:
> 
> 1. The expiry time is using the DBMS's clock for cleanup, what if the DBMS clock differs from the application server? It would be possible to use a timestamp generated from PHP for the `WHERE` clause. Shouldn't we avoid this where possible?
> 2. What about clock drift? If there are multiple servers spanning geographic regions should the developer be allowed to specify an acceptable drift (seconds) allowing cleanup to happen slightly after expiry? e.g. `MongoDbStore::createTtlIndex(int $expireAfterSeconds = 0)` The `$expireAfterSeconds` allows the developer / sysadmin to specify acceptable delay between the expiry and deletion.
> 
> I did also have a concern about people who run servers on different timezones however I'm not sure if we care about these people... :-p

This is a good question. Looks like @jderusse is solving for the problem by expecting clients to give a higher TTL from the beginning and it also assumes that the client and server locks are synchronized. I don't particularly see any problem with you generating a timestamp from PHP side to solve this problem.

Reply:
Assuming we are going to suggest a higher TTL to account for clock drift I have implemented automatic mongodb TTL Index creation. I did this using a probability modifier like php session / PDO store however due to the incompatibility with MongoDB prior to 2.2 I have also implemented mongodb version detection and reverted to a delete expired documents for older servers. I have still allowed for a developer to set their probability to `0.0` (effectively disabling the automatic TTL creation) in case they wish to manually manage the creation of their TTL Index. Let me know what you think?

Reply:
The current implementation of PdoStore always use the SGDB internal clock. This would avoid collision when 2 servers are not time synced. I suggest to the the same here.

From an other hand, the documentation recommend to take clock drift into account when diffining the TTL (https://github.com/symfony/symfony-docs/pull/9875/files#diff-1dc6462c0bc00fcdc7cba99d2f13e400R499)

Reply:
The equivalent of `SQL: NOW()` wasn't introduced until MongoDB 2.6 for `UPDATE`: https://jira.mongodb.org/browse/SERVER-10911

This does not work for `INSERT`: https://jira.mongodb.org/browse/SERVER-13695

There is no equivalent of `DATE_ADD()`.

I looked into doing it this way first because I have concerns about PHP `$ttl` expiry not being strictly adhered to (due to clock drift) however it's not possible in MongoDB. In light of this I think it's down to setting a higher `$initialTtl / $ttl` if a developer needs to account for high drift.

Reply:
@jderusse is the current implementation ok with you since it's not possible to do all time from the DB's clock? I'll be sure to document this on https://github.com/symfony/symfony-docs/pull/9807

Reply:
Given mongoDb don't provide methods to manipulate dates on upsert. I'm fine with using the APP date, and warning the user in the documentation.